modules/basegl/include/modules/basegl/processors/instancerenderer.h

```diff
@@ -0,0 +1,117 @@
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ *********************************************************************************/
+#pragma once
+
+#include <modules/basegl/baseglmoduledefine.h>
+
+#include <inviwo/core/processors/processor.h>
+#include <inviwo/core/interaction/cameratrackball.h>
+#include <inviwo/core/ports/meshport.h>
+#include <inviwo/core/ports/imageport.h>
+#include <inviwo/core/properties/boolproperty.h>
+#include <inviwo/core/properties/buttonproperty.h>
+#include <inviwo/core/properties/cameraproperty.h>
+#include <inviwo/core/properties/ordinalproperty.h>
+#include <inviwo/core/properties/compositeproperty.h>
+#include <inviwo/core/properties/simplelightingproperty.h>
+#include <inviwo/core/properties/listproperty.h>
+#include <inviwo/core/properties/stringproperty.h>
+#include <inviwo/core/rendering/meshdrawer.h>
+#include <inviwo/core/util/glmvec.h>
+#include <inviwo/core/util/stdextensions.h>
+
+#include <modules/opengl/shader/shader.h>
+#include <modules/opengl/shader/shadersegment.h>
+
+#include <vector>
+#include <tuple>
+
+namespace inviwo {
+
+class InstanceRenderer;
+
+namespace detail {
+
+struct IVW_MODULE_BASEGL_API DynPortManager {
+    DynPortManager(InstanceRenderer* theRenderer, std::unique_ptr<Inport> aPort,
+                   std::function<size_t()> aSize, std::function<void(Shader&, size_t)> aSet,
+                   std::function<void(ShaderObject&)> aAddUniform);
+
+    DynPortManager(const DynPortManager&) = delete;
+    DynPortManager& operator=(const DynPortManager&) = delete;
+    DynPortManager(DynPortManager&&);
+    DynPortManager& operator=(DynPortManager&&);
+    ~DynPortManager();
+
+    InstanceRenderer* renderer;
+    std::unique_ptr<Inport> port;
+    std::function<size_t()> size;
+    std::function<void(Shader&, size_t)> set;
+    std::function<void(ShaderObject&)> addUniform;
+};
+
+};  // namespace detail
+
+class IVW_MODULE_BASEGL_API InstanceRenderer : public Processor, public PropertyOwnerObserver {
```

Comment:
Document

If I understand the code correctly it renders the mesh input once per added port/buffer?
I would expect an instance renderer to render the mesh using `glDrawArrayInstanced `or `glDrawElementsInstanced`
https://learnopengl.com/Advanced-OpenGL/Instancing